Lambda School Data Science

*Unit 1, Sprint 1, Module 3*

---

# Join and Reshape datasets

Objectives
- concatenate data with pandas
- merge data with pandas
-  understand tidy data formatting
-  melt and pivot data with pandas

Links
- [Pandas Cheat Sheet](https://github.com/pandas-dev/pandas/blob/master/doc/cheatsheet/Pandas_Cheat_Sheet.pdf)
- [Tidy Data](https://en.wikipedia.org/wiki/Tidy_data)
  - Combine Data Sets: Standard Joins
  - Tidy Data
  - Reshaping Data
- Python Data Science Handbook
  - [Chapter 3.6](https://jakevdp.github.io/PythonDataScienceHandbook/03.06-concat-and-append.html), Combining Datasets: Concat and Append
  - [Chapter 3.7](https://jakevdp.github.io/PythonDataScienceHandbook/03.07-merge-and-join.html), Combining Datasets: Merge and Join
  - [Chapter 3.8](https://jakevdp.github.io/PythonDataScienceHandbook/03.08-aggregation-and-grouping.html), Aggregation and Grouping
  - [Chapter 3.9](https://jakevdp.github.io/PythonDataScienceHandbook/03.09-pivot-tables.html), Pivot Tables
  
Reference
- Pandas Documentation: [Reshaping and Pivot Tables](https://pandas.pydata.org/pandas-docs/stable/reshaping.html)
- Modern Pandas, Part 5: [Tidy Data](https://tomaugspurger.github.io/modern-5-tidy.html)

###Uploading the data

In [1]:
!wget https://s3.amazonaws.com/instacart-datasets/instacart_online_grocery_shopping_2017_05_01.tar.gz

--2020-01-28 22:52:04--  https://s3.amazonaws.com/instacart-datasets/instacart_online_grocery_shopping_2017_05_01.tar.gz
Resolving s3.amazonaws.com (s3.amazonaws.com)... 52.216.164.173
Connecting to s3.amazonaws.com (s3.amazonaws.com)|52.216.164.173|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 205548478 (196M) [application/x-gzip]
Saving to: ‘instacart_online_grocery_shopping_2017_05_01.tar.gz’

instacart_online_gr 100%[===================>] 196.03M  43.6MB/s    in 4.9s    

2020-01-28 22:52:10 (39.9 MB/s) - ‘instacart_online_grocery_shopping_2017_05_01.tar.gz’ saved [205548478/205548478]



In [2]:
!tar --gunzip --extract --verbose --file=instacart_online_grocery_shopping_2017_05_01.tar.gz

instacart_2017_05_01/
instacart_2017_05_01/._aisles.csv
instacart_2017_05_01/aisles.csv
instacart_2017_05_01/._departments.csv
instacart_2017_05_01/departments.csv
instacart_2017_05_01/._order_products__prior.csv
instacart_2017_05_01/order_products__prior.csv
instacart_2017_05_01/._order_products__train.csv
instacart_2017_05_01/order_products__train.csv
instacart_2017_05_01/._orders.csv
instacart_2017_05_01/orders.csv
instacart_2017_05_01/._products.csv
instacart_2017_05_01/products.csv


In [3]:
%cd instacart_2017_05_01

/content/instacart_2017_05_01


In [4]:
!ls -lh *.csv

-rw-r--r-- 1 502 staff 2.6K May  2  2017 aisles.csv
-rw-r--r-- 1 502 staff  270 May  2  2017 departments.csv
-rw-r--r-- 1 502 staff 551M May  2  2017 order_products__prior.csv
-rw-r--r-- 1 502 staff  24M May  2  2017 order_products__train.csv
-rw-r--r-- 1 502 staff 104M May  2  2017 orders.csv
-rw-r--r-- 1 502 staff 2.1M May  2  2017 products.csv


## Assignment

## Join Data Practice

These are the top 10 most frequently ordered products. How many times was each ordered? 

1. Banana
2. Bag of Organic Bananas
3. Organic Strawberries
4. Organic Baby Spinach 
5. Organic Hass Avocado
6. Organic Avocado
7. Large Lemon 
8. Strawberries
9. Limes 
10. Organic Whole Milk

First, write down which columns you need and which dataframes have them.

Next, merge these into a single dataframe.

Then, use pandas functions from the previous lesson to get the counts of the top 10 most frequently ordered products.

In [0]:
##### YOUR CODE HERE #####
import pandas as pd
import numpy as np

##Checking the datasets

In [6]:
aisles=pd.read_csv('aisles.csv')
print(aisles.shape)
aisles.sample(5)

(134, 2)


aisle_id          aisle
131       132         beauty
36         37  ice cream ice
19         20   oral hygiene
18         19  oils vinegars
130       131      dry pasta

In [7]:
departments=pd.read_csv('departments.csv')
print(departments.shape)
departments.sample(5)

(21, 2)


department_id     department
4               5        alcohol
12             13         pantry
7               8           pets
17             18         babies
10             11  personal care

In [8]:
order_products__prior=pd.read_csv('order_products__prior.csv')
print(order_products__prior.shape)
order_products__prior.sample(5)

(32434489, 4)


order_id  product_id  add_to_cart_order  reordered
12020080   1268816       37067                 13          1
26001464   2741997        3599                  8          1
32286915   3405479       35520                  1          0
26780452   2824036       20989                 18          0
20520595   2164652       38772                 18          0

In [9]:
order_products__train=pd.read_csv('order_products__train.csv')
print(order_products__train.shape)
order_products__train.sample(5)

(1384617, 4)


order_id  product_id  add_to_cart_order  reordered
767663    1889778       36793                  6          0
470069    1151916       30352                  4          1
1250586   3090157       27555                  9          0
1036085   2561242       23288                  3          0
560750    1378011       19660                 14          1

In [10]:
orders=pd.read_csv('orders.csv')
print(orders.shape)
orders.sample(5)

(3421083, 7)


order_id  user_id  ... order_hour_of_day  days_since_prior_order
2210865   2272892   133179  ...                14                     9.0
246371     551674    14896  ...                 7                     5.0
3415031   2036832   205833  ...                12                    16.0
1569006   3255142    94261  ...                 8                     4.0
1960542    299660   117718  ...                12                     1.0

[5 rows x 7 columns]

In [11]:
products=pd.read_csv('products.csv')
print(products.shape)
products.sample(5)

(49688, 4)


product_id                      product_name  aisle_id  department_id
4510         4511  Multi-Purpose Un-Petroleum Jelly       102             18
31458       31459        Sprouted Seven-Grain Bread       112              3
2915         2916  Gluten Free Seeds & Grains Bread       112              3
46746       46747        Dill & Sour Cream Plentils       107             19
45609       45610                Sweet Potato Fries       123              4

##Joining the datasets

In [12]:
#For merging the datasets I'll need the following columns:
# 1. 'product_id' and 'product_name' in "products"
# 2. 'add_to_cart_order' and 'reorderd in the df
#     that i'm about to create by concatenating
#     'order_products__prior' and 'order_products__train'
df=pd.concat([order_products__prior,order_products__train], axis=0)
# I didn't need to type 'axis=0' since it's considered to be there by default
print(df.shape)
df.head(10)

(33819106, 4)


order_id  product_id  add_to_cart_order  reordered
0         2       33120                  1          1
1         2       28985                  2          1
2         2        9327                  3          0
3         2       45918                  4          1
4         2       30035                  5          0
5         2       17794                  6          1
6         2       40141                  7          1
7         2        1819                  8          1
8         2       43668                  9          0
9         3       33754                  1          1

In [13]:
# Now let's merge this big dataframe with 'products'
df=pd.merge(df, products, on='product_id', how='inner')
print(df.shape)
df.head()

(33819106, 7)


order_id  product_id  ...  aisle_id  department_id
0         2       33120  ...        86             16
1        26       33120  ...        86             16
2       120       33120  ...        86             16
3       327       33120  ...        86             16
4       390       33120  ...        86             16

[5 rows x 7 columns]

In [14]:
# Looks like it worked. Yay! ^-^
# Now let's remove the columns that we don't need
df=df.drop(['aisle_id','department_id'],axis=1)
# 'axis=1' tells the computer that we need vertical set (column, not row)
print(df.shape)
df.head()

(33819106, 5)


order_id  product_id  add_to_cart_order  reordered        product_name
0         2       33120                  1          1  Organic Egg Whites
1        26       33120                  5          0  Organic Egg Whites
2       120       33120                 13          0  Organic Egg Whites
3       327       33120                  5          1  Organic Egg Whites
4       390       33120                 28          1  Organic Egg Whites

In [15]:
# Time to count them
df['product_name'].value_counts()[:10]


Banana                    491291
Bag of Organic Bananas    394930
Organic Strawberries      275577
Organic Baby Spinach      251705
Organic Hass Avocado      220877
Organic Avocado           184224
Large Lemon               160792
Strawberries              149445
Limes                     146660
Organic Whole Milk        142813
Name: product_name, dtype: int64

## Reshape Data Section

- Replicate the lesson code. Remember, if you haven't followed along typing out what we did during lecture, do that now to get more repetition with the syntax.
- Take table 2 (the transpose of table 1) and practice changing it into Tidy format and back again. You should not use the transpose operation anywhere in this code:
 - Table 2 --> Tidy
 - Tidy --> Table 2
- Load seaborn's `flights` dataset by running the cell below. Then create a pivot table showing the number of passengers by month and year. Use year for the index and month for the columns. You've done it right if you get 112 passengers for January 1949 and 432 passengers for December 1960.

In [16]:
import pandas as pd
import numpy as np

table1 = pd.DataFrame(
    [[np.nan, 2],
     [16,    11], 
     [3,      1]],
    index=['John Smith', 'Jane Doe', 'Mary Johnson'], 
    columns=['treatmenta', 'treatmentb'])

table2 = table1.T

table2

John Smith  Jane Doe  Mary Johnson
treatmenta         NaN      16.0           3.0
treatmentb         2.0      11.0           1.0

In [0]:
# table2 -> tidy

In [0]:
# tidy -> table2

In [19]:
flights = sns.load_dataset('flights')

NameError: ignored

In [0]:
# Flights Pivot Table

## Join Data Stretch Challenge

The [Instacart blog post](https://tech.instacart.com/3-million-instacart-orders-open-sourced-d40d29ead6f2) has a visualization of "**Popular products** purchased earliest in the day (green) and latest in the day (red)." 

The post says,

> "We can also see the time of day that users purchase specific products.

> Healthier snacks and staples tend to be purchased earlier in the day, whereas ice cream (especially Half Baked and The Tonight Dough) are far more popular when customers are ordering in the evening.

> **In fact, of the top 25 latest ordered products, the first 24 are ice cream! The last one, of course, is a frozen pizza.**"

Your challenge is to reproduce the list of the top 25 latest ordered popular products.

We'll define "popular products" as products with more than 2,900 orders.



In [0]:
##### YOUR CODE HERE #####

## Reshape Data Stretch Challenge

_Try whatever sounds most interesting to you!_

- Replicate more of Instacart's visualization showing "Hour of Day Ordered" vs "Percent of Orders by Product"
- Replicate parts of the other visualization from [Instacart's blog post](https://tech.instacart.com/3-million-instacart-orders-open-sourced-d40d29ead6f2), showing "Number of Purchases" vs "Percent Reorder Purchases"
- Get the most recent order for each user in Instacart's dataset. This is a useful baseline when [predicting a user's next order](https://www.kaggle.com/c/instacart-market-basket-analysis)
- Replicate parts of the blog post linked at the top of this notebook: [Modern Pandas, Part 5: Tidy Data](https://tomaugspurger.github.io/modern-5-tidy.html)

In [0]:
##### YOUR CODE HERE #####